Baseline helm chart

ctab opened the pull request:
Using https://github.com/open-policy-agent/gatekeeper/pull/161 as a starting point (thanks to @sylus for the initial commit and setting up the structure), I restructured to (a) use kustomize instead of yq for substitution, and (b) use our existing deployment pipeline for the helm template so we wouldn't have to maintain two separate collections of YAML files.

I also removed some of the substitutable values contained in the other PR, mainly that as a singleton app we shouldn't need to support name subsitution, and the RBAC/frameworks CRDS should always be installed.

(Issue: https://github.com/open-policy-agent/gatekeeper/issues/5)

maxsmythe commented on Makefile:
```diff
@@ -94,9 +96,29 @@ deploy: manifests
+	@bash -c 'for x in vendor/github.com/open-policy-agent/frameworks/constraint/deploy/*.yaml ; do echo --- >> deploy/gatekeeper.yaml ; cat $${x} >> deploy/gatekeeper.yaml ; done'
+	@bash -c 'for x in vendor/github.com/open-policy-agent/frameworks/constraint/deploy/*.yaml ; do echo --- >> ${CHART_PATH}/helm-modifications/_temp.yaml ; cat $${x} >> ${CHART_PATH}/helm-modifications/_temp.yaml ; done'
+	@kustomize build ${CHART_PATH}/helm-modifications -o ${CHART_PATH}/templates/gatekeeper.yaml
+	@sed -i -E "s/STATEFUL_SET_RESOURCES/\
```
I'm not in love with the `sed` thing. That being said, I'm not sure I can come up with a better solution that doesn't add more dependencies to the build server.

We should:

   * Add a likely-to-be-unique prefix to these anchor values to help avoid collision, ex.: HELMSUBST_STATEFUL_SET_RESSOURCES
  * create a bash script that lives in the `helm-modifications` dir so that all helm-related stuff is located close to each other and to avoid cluttering the makefile

maxsmythe commented on Makefile:
```diff
@@ -161,9 +183,19 @@ docker-push:
+release-manifest: generate-helm-chart
 	@sed -i'' -e 's@image: $(REPOSITORY):.*@image: $(REPOSITORY):'"$(NEWVERSION)"'@' ./config/manager/manager.yaml ./deploy/gatekeeper.yaml
 
+generate-helm-chart:
```
This name is currently misleading. I suggest you move the helm-specific logic in `manifests` here and make `manifests` a dependency of `generate-helm-chart`.

ctab replied:
Since this one (which updates the charts, not the templates) requires a version to be supplied I thought it would be better to keep it separate from the template generation. If we put it all together then `make manifests` would now require a version to be passed in. (Unless I'm misunderstanding what you are suggesting.)

maxsmythe replied:
We should move the version-updating stuff to `release-manifest`

I think you're right that there's a problem with dependencies in my original suggestion. I think we can just remove this rule altogether once `release-manifest` is updated

ctab replied:
Sure thing, I've combined it all and removed the extra rule.

maxsmythe commented on README.md:
```diff
@@ -68,6 +68,10 @@ Currently the most reliable way of installing Gatekeeper is to build and install
 
+#### Deploying via Helm ####
+
+A basic Helm template exists in `hack/chart/gatekeeper-operator`. If you have Helm installed and Tiller initialized on your cluster you can deploy via `helm install hack/chart/gatekeeper-operator/`
```
We should definitely move the chart out of the `hack` directory if it's user-facing.

maxsmythe commented on Makefile:
```diff
@@ -1,3 +1,5 @@
+CHART_PATH := $(PWD)/hack/chart/gatekeeper-operator
```
Also, we can drop `$(PWD)/` the PWD var is evaluating to the empty string on the build system, which is breaking things

ctab replied:
I've moved everything out of `hack`. Assuming that was just a place to put it while it was in progress but it should be fine to move it out now.

sozercan replied:
charts usually go in `/chart` or `/charts`

maxsmythe commented on Makefile:
```diff
@@ -1,3 +1,5 @@
+CHART_PATH := $(PWD)/hack/chart/gatekeeper-operator
+CONFIG_PATH := $(PWD)/config
```
unused variable, please remove

maxsmythe commented:
Also, we may want to test that the helm chat can install Gatekeeper successfully... maybe run the e2e tests against them?

sozercan commented on chart/gatekeeper-operator/templates/gatekeeper.yaml:
```diff
@@ -0,0 +1,517 @@
+    spec:
+      containers:
+      - args:
+        - --auditInterval=30
```
can we have `auditInterval` and `constraintViolationsLimit` as variables? (defaulting to their default values)

ctab replied:
Yes @sozercan good call, thanks for the suggestion.

ctab commented:
@maxsmythe thanks for the review. I left you one comment up there and addressed the other points you mentioned (moved out of hack directory, moved logic from Makefile to a separate `.sh` file in the chart directory, cleaning up unused variable and unneeded ${PWD}).

sozercan commented on chart/gatekeeper-operator/helm-modifications/helm-modifications.yaml:
```diff
@@ -0,0 +1,53 @@
+metadata:
+  name: configs.config.gatekeeper.sh
+  annotations:
+    helm.sh/hook: crd-install
```
`crd-install` hook is not supported in Helm 3. Can we use the workaround at https://github.com/helm/helm/pull/6243?

excavador replied:
Actual stable version of helm is Helm 2, and it will be so long time (until helm 3 would be stable).
Until Helm 3 released we still need helm chart.

Please, please, please, keep this hook.

ctab replied:
Agreed, for this reason I prefer to leave this as is and specify that this is for Helm v2, and leave v3 for another day.

ritazh commented on README.md:
```diff
@@ -68,6 +68,10 @@ Currently the most reliable way of installing Gatekeeper is to build and install
+
+A basic Helm template exists in `chart/gatekeeper-operator`. If you have Helm installed and Tiller initialized on your cluster you can deploy via `helm install chart/gatekeeper-operator/`
+
 ### Uninstallation
```
We need to have some instructions for `Uninstalling via Helm`.

ctab replied:
:+1: I added a section under Uninstallation.

ritazh commented on Makefile:
```diff
@@ -161,9 +162,19 @@ docker-push:
+	@echo "Updating chart images tag to: ${NEWVERSION}"
+	@sed -i -E "s/image: (.*):.*/image: \1:${NEWVERSION}/" ${CHART_PATH}/values.yaml
+	@sed -i -E "s/sidecarImage: (.*):.*/sidecarImage: \1:${NEWVERSION}/" ${CHART_PATH}/values.yaml
+	@sed -i -E "s/	image: (.*): (.*):.*/	image: \1:${NEWVERSION}/" ${CHART_PATH}/values.yaml
```
maybe i missed this in values.yaml, but where is `sidecarImage` and why is `image` updated 2x?

ctab replied:
You're right -- not needed, will remove. Thanks!

ritazh commented on README.md:
```diff
@@ -68,6 +68,10 @@ Currently the most reliable way of installing Gatekeeper is to build and install
 
+#### Deploying via Helm ####
+
+A basic Helm template exists in `chart/gatekeeper-operator`. If you have Helm installed and Tiller initialized on your cluster you can deploy via `helm install chart/gatekeeper-operator/`
```
Lets decide on if we want to use helm v2 vs helm v3 as the instructions and experience are different.

ctab replied:
Can we stick with v2 as that is currently the latest stable version (and what this all works with)? I can note this in the README.

ArgTang replied:
helm v3 is released, so it would be best to support both

ritazh commented on chart/gatekeeper-operator/generate_helm_template.sh:
```diff
@@ -0,0 +1,24 @@
+sed -i "s/HELMSUBST_VALUES_AUDITINTERVAL_PLACEHOLDER/{{ .Values.auditInterval }}/g" templates/gatekeeper.yaml
+sed -i "s/HELMSUBST_VALUES_CONSTRAINTVIOLATIONSLIMIT_PLACEHOLDER/{{ .Values.constraintViolationsLimit }}/g" templates/gatekeeper.yaml
+rm ./helm-modifications/_temp.yaml
+echo "Helm template created under 'chart/gatekeeper-operator/templates'"
```
nit: line break

ritazh commented on chart/gatekeeper-operator/generate_helm_template.sh:
```diff
@@ -0,0 +1,24 @@
+\n    {{- end }}/" templates/gatekeeper.yaml
+sed -i "s/HELMSUBST_VALUES_REPLICAS_PLACEHOLDER/{{ .Values.Replicas }}/g" templates/gatekeeper.yaml
+sed -i "s/HELMSUBST_VALUES_AUDITINTERVAL_PLACEHOLDER/{{ .Values.auditInterval }}/g" templates/gatekeeper.yaml
+sed -i "s/HELMSUBST_VALUES_CONSTRAINTVIOLATIONSLIMIT_PLACEHOLDER/{{ .Values.constraintViolationsLimit }}/g" templates/gatekeeper.yaml
```
Instead of doing these seds, can we update helm-modifications.yaml to use the parameters from values.yaml? 
Example:  
```
replicas:  {{ .Value.Replicas}}
```

ctab replied:
I hate doing it this way but kustomize was giving me an error if I put ``{{ .Values.Replicas }}`` without quotes around it (which then becomes an invalid value because it's a string rather than an integer):
```
Error: builtin PatchStrategicMergeTransformer config: [112 97 116 104 115 58 10 45 32 104 101 108 109 45 109 111 100 105 102 105 99 97 116 105 111 110 115 46 121 97 109 108 10]: YAML file [helm-modifications.yaml] encounters a format error.
error converting YAML to JSON: yaml: invalid map key: map[interface {}]interface {}{".Values.replicas":interface {}(nil)}
```
If you happen to know a way around that please let me know. Otherwise this was a messy workaround.

In any event it turns out this issue doesn't affect the other arg parameters so I removed those sed lines.

ritazh commented on chart/gatekeeper-operator/values.yaml:
```diff
@@ -0,0 +1,10 @@
+replicas: 1
+auditInterval: 60
+constraintViolationsLimit: 20
+image: quay.io/open-policy-agent/gatekeeper:v3.0.4-beta.2
```
can we consider something like this for the image so that we only need to update the tag when we have a release.

```yaml
image:
  repository: quay.io/open-policy-agent/gatekeeper
  tag: v3.0.4-beta.2
  pullPolicy: IfNotPresent
```

ctab replied:
Sure, good idea

maxsmythe commented on .travis.yml:
```diff
@@ -33,6 +33,29 @@ jobs:
+        - make e2e-helm-deploy HELM_REPO=gatekeeper-e2e HELM_RELEASE=latest
+        - make test-e2e
+        - echo -e '\n\n======= manager logs =======\n\n' && kubectl logs -n gatekeeper-system gatekeeper-controller-manager-0 manager
+      deploy:
```
We should have `deploy` steps for only one stage... likely not this one.

ctab replied:
Good catch - removed

maxsmythe commented on .travis.yml:
```diff
@@ -35,6 +35,14 @@ jobs:
             tags: true # only push release image tag for tags
+    - stage: "helm build, test"
+      script:
+        - make manager
```
We can drop the `make manager` line here, linting is taken care of elsewhere as well

maxsmythe commented:
This LGTM to me after the Travis step nit.

@ritazh Good to merge on your end?

ritazh commented on chart/gatekeeper-operator/generate_helm_template.sh:
```diff
@@ -0,0 +1,22 @@
+#!/bin/bash
+scriptdir="$(dirname "$0")"
+cd "$scriptdir"
+cp ./../../deploy/gatekeeper_kubebuilder_v2.yaml ${PWD}/helm-modifications/_temp.yaml
```
this should be gatekeeper.yaml now that we merged v2 changes.

ritazh commented on Makefile:
```diff
@@ -177,7 +188,17 @@ release:
+travis-dev-deploy: docker-login docker-build-ci docker-push-dev
+
+# Travis Release
+travis-release-deploy: docker-login docker-build-ci docker-push-release
```
nit: these travis targets were recently removed.

ritazh commented on README.md:
````diff
@@ -112,6 +121,13 @@ If you used `make` to deploy, then run the following to uninstall Gatekeeper:
+
+If you used `helm` to deploy, then run the following to uninstall Gatekeeper:
+```sh
+helm delete <release name>
````
nit: `helm delete <release name> --purge`

ritazh commented on chart/gatekeeper-operator/values.yaml:
```diff
@@ -0,0 +1,9 @@
+constraintViolationsLimit: 20
+image:
+  repository: quay.io/open-policy-agent/gatekeeper
+  release: v3.0.4-beta.2
```
let's update this to the latest release so it will work with the gatekeeper.yaml in the chart

ritazh commented on chart/gatekeeper-operator/Chart.yaml:
```diff
@@ -0,0 +1,10 @@
+name: gatekeeper-operator
+keywords:
+  - open policy agent
+version: v3.0.4-beta.2
```
lets update this to the latest release

ritazh commented on chart/gatekeeper-operator/Chart.yaml:
```diff
@@ -0,0 +1,10 @@
+home: https://github.com/open-policy-agent/gatekeeper
+sources:
+  - https://github.com/open-policy-agent/gatekeeper.git
+appVersion: v3.0.4-beta.2
```
lets update this to the latest release

ctab commented:
@ritazh thanks for catching all those things -- fixed.
